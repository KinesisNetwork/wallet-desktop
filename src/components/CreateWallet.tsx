import { Keypair } from 'js-kinesis-sdk'
import { kebabCase, startCase } from 'lodash'
import * as React from 'react'

import { InputField } from '@components/InputField'
import { formAlert } from '@helpers/alert'
import { InputError } from '@helpers/errors'
import { encryptWithPassword } from '@services/encryption'
import { CreateWalletForm, CreateWalletFormView as FormView, Wallet } from '@types'

export interface Props extends CreateWalletForm {
  activeView: FormView
  addWallet: (wallet: Wallet) => any
  handleChange: (field: keyof CreateWalletForm, newValue: string) => any
  changeFormView: (newView: FormView) => any
}

export const CreateWallet: React.SFC<Props> = props => (
  <div className="vertical-spaced has-text-centered">
    <h1 className="title-heading">ADD A NEW ACCOUNT</h1>
    <section className="section">
      {props.activeView === FormView.select && (
        <FormSelection changeFormView={props.changeFormView} />
      )}
      {props.activeView !== FormView.select && <WalletForm {...props} />}
    </section>
  </div>
)

CreateWallet.displayName = 'CreateWallet'

const FormSelection: React.SFC<Pick<Props, 'changeFormView'>> = ({ changeFormView }) => (
  <div className="buttons is-centered">
    <button className="button" onClick={() => changeFormView(FormView.generate)}>
      Generate New Account
    </button>
    <button className="button" onClick={() => changeFormView(FormView.import)}>
      Import Existing Account
    </button>
  </div>
)

export class WalletForm extends React.Component<Props> {
  displayName = 'WalletForm'

  constructor(props: Props) {
    super(props)
  }

  createNewWallet = async (ev: React.FormEvent<HTMLFormElement>) => {
    ev.preventDefault()
    try {
      this.validateProps()
      const { publicKey, privateKey } = this.generateKeyOrExtractFromProps()
      const encryptedPrivateKey = encryptWithPassword(privateKey, this.props.password)
      this.props.addWallet({
        encryptedPrivateKey,
        publicKey,
        accountName: this.props.accountName,
      })
    } catch (e) {
      formAlert(e.message, e.key)
    }
  }

  validateProps = (): void | never => {
    this.checkValidEntry('accountName')
    if (this.props.activeView === FormView.import) {
      this.checkValidEntry('privateKey')
    }
    this.checkValidEntry('password')
    this.checkValidPassword()
  }

  generateKeyOrExtractFromProps = () => {
    const keypair =
      this.props.activeView === FormView.import
        ? Keypair.fromSecret(this.props.privateKey)
        : Keypair.random()
    return { publicKey: keypair.publicKey(), privateKey: keypair.secret() }
  }

  checkValidEntry = (key: keyof CreateWalletForm) => {
    if (!this.props[key]) {
      throw new InputError(`${startCase(key)} is required`, `input-${kebabCase(key)}`)
    }
  }

  checkValidPassword = () => {
    const { password, passwordVerify } = this.props
    const lengthOfPassword = password.length
    if (lengthOfPassword < 12 || lengthOfPassword > 30) {
      throw new InputError(
        'Password needs to be between 12 and 30 characters long',
        `input-${kebabCase('password')}`,
      )
    }

    if (password !== passwordVerify) {
      throw new InputError(`Passwords don't match`, `input-${kebabCase('passwordVerify')}`)
    }
  }

  renderImportFields = () => (
    <InputField
      label="Private Key"
      value={this.props.privateKey}
      id="private-key"
      helpText="Add your private key"
      onChangeHandler={newValue => this.props.handleChange('privateKey', newValue)}
    />
  )

  render() {
    const {
      activeView,
      accountName,
      password,
      passwordVerify,
      changeFormView,
      handleChange,
    } = this.props
    const action = activeView === FormView.import ? 'Import' : 'Generate'
    return (
      <div>
        <div>
          <span className="icon is-large">
            <i className="far fa-user" />
          </span>
          <h1 className="title is-4 heading primary-font">{`${action} Account`}</h1>
        </div>
        <div className="columns is-centered">
          <form onSubmit={this.createNewWallet} className="column is-half">
            <InputField
              label="Account Name"
              value={accountName}
              id="account-name"
              helpText="Add an alias for your account"
              onChangeHandler={newValue => handleChange('accountName', newValue)}
            />
            {activeView === FormView.import && this.renderImportFields()}
            <InputField
              label="Account Password"
              value={password}
              type={'password'}
              id="password"
              helpText="Add a password for locking this account"
              onChangeHandler={newValue => handleChange('password', newValue)}
            />
            <InputField
              label="Repeat Password"
              value={passwordVerify}
              type={'password'}
              id="password-verify"
              onChangeHandler={newValue => handleChange('passwordVerify', newValue)}
            />
            <div className="field is-grouped">
              <div className="control is-expanded">
                <button className="button is-fullwidth" type="submit">{`${action} Account`}</button>
              </div>
              <div className="control">
                <button
                  className="button is-danger"
                  type="button"
                  onClick={() => changeFormView(FormView.select)}
                >
                  Back
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    )
  }
}
