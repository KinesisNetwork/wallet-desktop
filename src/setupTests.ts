import { configure } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

export default configure({ adapter: new Adapter() })
