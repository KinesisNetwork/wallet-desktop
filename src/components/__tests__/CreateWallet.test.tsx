import { shallow } from 'enzyme';
import * as React from 'react'

import { CreateWallet, WalletForm } from '@components/CreateWallet'
import '../../setupTests'

describe('CreateWallet', () => {
  it('renders correctly', () => {
    const wrapperWhenFormSelection = shallow(
      <CreateWallet
        accountName=''
        activeView={0}
        addWallet={() => null}
        changeFormView={() => null}
        handleChange={() => null}
        password=''
        passwordVerify=''
        privateKey=''
      />
    )
    const wrapperWhenWalletForm = shallow(
      <CreateWallet
        accountName=''
        activeView={1}
        addWallet={() => null}
        changeFormView={() => null}
        handleChange={() => null}
        password=''
        passwordVerify=''
        privateKey=''
      />
    )

    expect(wrapperWhenFormSelection).toMatchSnapshot()
    expect(wrapperWhenWalletForm).toMatchSnapshot()
  })

  it('should render a primary heading and a FormSelection component if FormView is selected', () => {
    const wrapper = shallow(
      <CreateWallet
        accountName=''
        activeView={0}
        addWallet={() => null}
        changeFormView={() => null}
        handleChange={() => null}
        password=''
        passwordVerify=''
        privateKey=''
      />
    )

    expect(wrapper.find('h1')).toHaveLength(1)
    expect(wrapper.find('h1').text()).toEqual('ADD A NEW ACCOUNT')
    expect(wrapper.find('FormSelection')).toHaveLength(1)
  })

  it('renders a WalletForm if selected as activeView', () => {
    const wrapper = shallow(
      <CreateWallet
        accountName=''
        activeView={1}
        addWallet={() => null}
        changeFormView={() => null}
        handleChange={() => null}
        password=''
        passwordVerify=''
        privateKey=''
      />
    )

    expect(wrapper.find('WalletForm')).toHaveLength(1)
  })
})

describe('WalletForm', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <WalletForm
        accountName=''
        activeView={1}
        addWallet={() => null}
        changeFormView={() => null}
        handleChange={() => null}
        password=''
        passwordVerify=''
        privateKey=''
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('renders a primary heading with an icon', () => {
    const wrapper = shallow(
      <WalletForm
        accountName=''
        activeView={1}
        addWallet={() => null}
        changeFormView={() => null}
        handleChange={() => null}
        password=''
        passwordVerify=''
        privateKey=''
      />
    )

    expect(wrapper.find('h1').text()).toEqual('Generate Account')
    expect(wrapper.find('.icon').children().hasClass('fa-user')).toBe(true)
  })

  it('displays a form with three InputForm components and two buttons', () => {
    const wrapper = shallow(
      <WalletForm
        accountName=''
        activeView={1}
        addWallet={() => null}
        changeFormView={() => null}
        handleChange={() => null}
        password=''
        passwordVerify=''
        privateKey=''
      />
    )

    expect(wrapper.find('form')).toHaveLength(1)
    expect(wrapper.find('form').children()).toHaveLength(4)
    expect(wrapper.find('InputField')).toHaveLength(3)
    expect(wrapper.find('button')).toHaveLength(2)
    expect(wrapper.find('button').at(0).text()).toEqual('Generate Account')
    expect(wrapper.find('button').at(1).text()).toEqual('Back')
  })

  it('calls createNewWallet function when submitted', () => {
    const mockCreateNewWallet = jest.fn()
    const wrapper = shallow(
      <WalletForm
        accountName='qwe'
        activeView={1}
        addWallet={mockCreateNewWallet}
        changeFormView={() => null}
        handleChange={() => null}
        password='asd'
        passwordVerify='asd'
        privateKey='zxc'
      />
    )

    const form = wrapper.find('form')
    form.simulate('submit', { preventDefault: () => null })
    expect(mockCreateNewWallet).toHaveBeenCalled()
  })

  it('calls changeFormView function when clicking on the Back button', () => {
    const mockChangeFormView = jest.fn()
    const wrapper = shallow(
      <WalletForm
        accountName='qwe'
        activeView={1}
        addWallet={() => null}
        changeFormView={mockChangeFormView}
        handleChange={() => null}
        password='asd'
        passwordVerify='asd'
        privateKey='zxc'
      />
    )

    const backButton = wrapper.find('.button.is-danger')
    backButton.simulate('click')
    expect(mockChangeFormView).toHaveBeenCalled()
  })
})
