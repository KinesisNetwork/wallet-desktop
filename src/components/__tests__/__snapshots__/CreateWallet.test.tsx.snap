// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateWallet renders correctly 1`] = `
<div
  className="vertical-spaced has-text-centered"
>
  <h1
    className="title-heading"
  >
    ADD A NEW ACCOUNT
  </h1>
  <section
    className="section"
  >
    <FormSelection
      changeFormView={[Function]}
    />
  </section>
</div>
`;

exports[`CreateWallet renders correctly 2`] = `
<div
  className="vertical-spaced has-text-centered"
>
  <h1
    className="title-heading"
  >
    ADD A NEW ACCOUNT
  </h1>
  <section
    className="section"
  >
    <WalletForm
      accountName=""
      activeView={1}
      addWallet={[Function]}
      changeFormView={[Function]}
      handleChange={[Function]}
      password=""
      passwordVerify=""
      privateKey=""
    />
  </section>
</div>
`;

exports[`WalletForm should render correctly 1`] = `
<div>
  <div>
    <span
      className="icon is-large"
    >
      <i
        className="far fa-user"
      />
    </span>
    <h1
      className="title is-4 heading primary-font"
    >
      Generate Account
    </h1>
  </div>
  <div
    className="columns is-centered"
  >
    <form
      className="column is-half"
      onSubmit={[Function]}
    >
      <InputField
        helpText="Add an alias for your account"
        id="account-name"
        label="Account Name"
        onChangeHandler={[Function]}
        value=""
      />
      <InputField
        helpText="Add a password for locking this account"
        id="password"
        label="Account Password"
        onChangeHandler={[Function]}
        type="password"
        value=""
      />
      <InputField
        id="password-verify"
        label="Repeat Password"
        onChangeHandler={[Function]}
        type="password"
        value=""
      />
      <div
        className="field is-grouped"
      >
        <div
          className="control is-expanded"
        >
          <button
            className="button is-fullwidth"
            type="submit"
          >
            Generate Account
          </button>
        </div>
        <div
          className="control"
        >
          <button
            className="button is-danger"
            onClick={[Function]}
            type="button"
          >
            Back
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
